File: codec/elixir/lib/encode.ex
defmodule DcgCodec.Encode do
  def encode(d, version \\ DcgCodec.Common.version()) do
    language_number = Map.get(d, "language") |> DcgCodec.Language.to_int()

    digi_eggs = Map.get(d, "digi-eggs", [])

    version_and_digi_egg_count =
      if 3 <= version && version <= 4 do
        Bitwise.bsl(version, 4)
        |> Bitwise.bor(Bitwise.bsl(language_number, 3))
        |> Bitwise.bor(Bitwise.band(length(digi_eggs), 0x0F))
      else
        Bitwise.bsl(version, 4)
        |> Bitwise.bor(Bitwise.band(length(digi_eggs), 0x0F))
      end

    icon = Map.get(d, "icon")
    deck_name = Map.get(d, "name", "")

    deck_name =
      if version >= 4 && icon do
        (icon
         |> String.slice(0, 8)
         |> String.pad_trailing(8, " ")) <> deck_name
      else
        deck_name
      end
      |> String.trim()

    deck_name_bytes = deck_name |> binary_part(0, min(0x3F, byte_size(deck_name)))

    deck_name_length_byte =
      if version >= 5 do
        Bitwise.bsl(language_number, 6)
        |> Bitwise.bor(byte_size(deck_name_bytes))
      else
        byte_size(deck_name_bytes)
      end

    deck_bytes = <<>> <> <<deck_name_length_byte>>

    sideboard = Map.get(d, "sideboard", [])

    deck_bytes =
      if version >= 2 do
        sideboard_count = length(sideboard)

        sideboard_count =
          if version >= 4 && icon do
            Bitwise.bor(sideboard_count, 0x80)
          else
            sideboard_count
          end

        deck_bytes <> <<sideboard_count>>
      else
        deck_bytes
      end

    deck = Map.get(d, "deck", [])

    deck_bytes =
      []
      |> Enum.concat(group_cards(digi_eggs))
      |> Enum.concat(group_cards(deck))
      |> Enum.concat(group_cards(sideboard))
      |> List.foldl(deck_bytes, fn {{card_set, card_number_padding}, cards}, acc ->
        deck_bytes =
          case version do
            0 ->
              acc <> String.pad_trailing(card_set, 4, " ")

            _ ->
              card_set_chars = String.graphemes(card_set)

              card_set_bytes =
                List.foldl(card_set_chars, <<>>, fn char, acc2 ->
                  {base_36_char, _} = Integer.parse(char, 36)

                  result =
                    if byte_size(acc2) + 1 < length(card_set_chars) do
                      Bitwise.bor(base_36_char, 0x80)
                    else
                      base_36_char
                    end

                  acc2 <> <<result>>
                end)

              acc <> card_set_bytes
          end

        deck_bytes =
          if version < 2 do
            x =
              Bitwise.bsl(card_number_padding - 1, 6)
              |> Bitwise.bor(length(cards))

            deck_bytes <> <<x>>
          else
            x =
              Bitwise.bsl(card_number_padding - 1, 6)
              |> Bitwise.bor(bits_with_carry(length(cards), 6))

            (deck_bytes <> <<x>>)
            |> append_rest_to_deck_bytes(length(cards), 6)
          end

        prev_card_number = 0

        {deck_bytes, _} =
          List.foldl(cards, {deck_bytes, prev_card_number}, fn card, acc ->
            card_count = Map.get(card, "count")
            card_number = Map.get(card, "number")
            {deck_bytes, prev_card_number} = acc
            [_, card_set_number] = String.split(card_number, "-")
            {card_set_number, _} = Integer.parse(card_set_number, 10)
            card_number_offset = card_set_number - prev_card_number
            prev_card_number = card_set_number
            parallel_id = Map.get(card, "parallel-id", 0)

            deck_bytes =
              if version == 0 do
                x =
                  Bitwise.bsl(card_count - 1, 6)
                  |> Bitwise.bor(Bitwise.bsl(parallel_id, 3))
                  |> Bitwise.bor(bits_with_carry(card_number_offset, 3))

                (deck_bytes <> <<x>>)
                |> append_rest_to_deck_bytes(card_number_offset, 3)
              else
                card_number_offset_start =
                  Bitwise.bor(Bitwise.bsl(parallel_id, 5), bits_with_carry(card_number_offset, 5))

                (deck_bytes <> <<card_count - 1>> <> <<card_number_offset_start>>)
                |> append_rest_to_deck_bytes(card_number_offset, 5)
              end

            {deck_bytes, prev_card_number}
          end)

        deck_bytes
      end)

    bytes_to_checksum = binary_part(deck_bytes, 1, byte_size(deck_bytes) - 1)

    computed_checksum =
      DcgCodec.Common.compute_checksum(bytes_to_checksum, byte_size(bytes_to_checksum))

    checksum = <<computed_checksum>>
    deck_bytes = <<version_and_digi_egg_count>> <> checksum <> deck_bytes <> deck_name_bytes

    DcgCodec.Common.prefix() <> Base.url_encode64(deck_bytes, padding: false)
  end

  defp bits_with_carry(value, bits) do
    limit_bit = Bitwise.bsl(1, bits - 1)
    result = Bitwise.band(value, limit_bit - 1)

    if value >= limit_bit do
      Bitwise.bor(result, limit_bit)
    else
      result
    end
  end

  defp append_rest_to_deck_bytes(deck_bytes, value, already_written_bits) do
    remaining_value = Bitwise.bsr(value, already_written_bits - 1)

    if remaining_value > 0 do
      x = bits_with_carry(remaining_value, 8)
      append_rest_to_deck_bytes(deck_bytes <> <<x>>, remaining_value, 8)
    else
      deck_bytes
    end
  end

  defp group_cards(cards) do
    Enum.group_by(cards, fn card ->
      card_number = Map.get(card, "number", "")
      [card_set, card_set_number] = String.split(card_number, "-")
      {card_set, String.length(card_set_number)}
    end)
    |> Enum.to_list()
    |> List.foldl([], fn group, acc ->
      {k, v} = group

      v =
        Enum.sort_by(v, fn card ->
          [Map.get(card, "number"), Map.get(card, "parallel-id", 0)]
        end)

      Enum.concat(acc, [{k, v}])
    end)
    |> Enum.sort_by(fn {{card_set, _}, _} ->
      card_set
    end)
  end
end

File: codec/typescript/src/codec/encode.ts
import {
  VERSION,
  PREFIX,
  Card,
  Deck,
  Language,
  char_to_base36,
  computeChecksum,
} from "./common";

async function bufferToBase64(buffer: Uint8Array): Promise<string> {
  return await new Promise((resolve, reject) => {
    const reader = Object.assign(new FileReader(), {
      onload: () => {
        const s = reader.result as string;
        resolve(s.substring(s.indexOf(",") + 1));
      },
      onerror: () => reject(reader.error),
    });
    reader.readAsDataURL(
      new File([buffer], "", { type: "application/octet-stream" }),
    );
  });
}

function bits_with_carry(value: number, bits: number): number {
  const limit_bit = 1 << (bits - 1);
  let result = value & (limit_bit - 1);
  if (value >= limit_bit) {
    result |= limit_bit;
  }
  return result;
}

function append_rest_to_deck_bytes(
  deckBytes: Array<number>,
  value: number,
  already_written_bits: number,
) {
  let remaining_value = value >> (already_written_bits - 1);
  while (remaining_value > 0) {
    deckBytes.push(bits_with_carry(remaining_value, 8));
    remaining_value = remaining_value >> 7;
  }
}

function group_cards(d: Array<Card>): Array<[string, number, Array<Card>]> {
  d.sort(
    (a, b) => a.number.localeCompare(b.number) || a.parallel_id - b.parallel_id,
  );
  let result = Object.groupBy(d, ({ number }) => {
    const [card_set, card_number] = number.split("-");
    return [card_set, card_number.length].toString();
  });
  return Object.entries(result).map(([k, v]) => {
    const [card_set, padding] = k.split(",");
    return [card_set, parseInt(padding), v];
  });
}

export async function encode(
  deck: Deck,
  version: number = VERSION,
): Promise<string> {
  let deckBytes: number[] = new Array();
  const textEncoder = new TextEncoder();
  const language_number = Language[Language[deck.language]] ?? 1;

  const version_and_digi_egg_count =
    3 <= version && version <= 4
      ? (version << 4) | (language_number << 3) | (deck.digi_eggs.length & 0x0f)
      : (version << 4) | (deck.digi_eggs.length & 0x0f);

  if (version >= 4 && !!deck.icon) {
    deck.name = deck.icon.padEnd(8) + deck.name;
  }

  let deck_name_bytes = textEncoder.encode(deck.name);
  deck_name_bytes = deck_name_bytes.slice(0, 0x3f);
  deck.name = new TextDecoder().decode(deck_name_bytes);
  let name_length = deck_name_bytes.length;
  if (version >= 5) {
    name_length = (language_number << 6) | name_length;
  }

  deckBytes.push(version_and_digi_egg_count);
  deckBytes.push(0); // checksum placeholder
  deckBytes.push(name_length);
  const HEADER_SIZE = deckBytes.length;

  if (version >= 2) {
    let sideboard_size = deck?.sideboard?.length ?? 0;
    if (version >= 4 && !!deck.icon) {
      sideboard_size |= 0x80;
    }
    deckBytes.push(sideboard_size);
  }

  let grouped_decks = new Array();
  grouped_decks = grouped_decks.concat(group_cards(deck?.digi_eggs ?? []));
  grouped_decks = grouped_decks.concat(group_cards(deck?.deck ?? []));
  grouped_decks = grouped_decks.concat(group_cards(deck?.sideboard ?? []));

  for (const [card_set, card_number_padding, grouped_cards] of grouped_decks) {
    // Encode card_set
    if (version == 0) {
      // Use 4 characters/bytes to store card sets.
      deckBytes = deckBytes.concat(
        Array.from(textEncoder.encode(card_set.padEnd(4, " "))),
      );
    } else {
      // Encode each character of card-set in Base36.
      // Use 8th bit as continue bit. If 0, reached end.
      for (let i = 0; i < card_set.length; i++) {
        const c = card_set[i];
        let base36_char = char_to_base36.get(c);
        if (!!card_set[parseInt(i.toString()) + 1]) {
          base36_char |= 0x80;
        }
        deckBytes.push(base36_char);
      }
    }
    // 2 bits for card number zero padding (zero padding stored as 0 indexed)
    // 6 bits for initial count offset of cards in a card group
    if (version < 2) {
      deckBytes.push(((card_number_padding - 1) << 6) | grouped_cards.length);
    } else {
      deckBytes.push(
        ((card_number_padding - 1) << 6) |
          bits_with_carry(grouped_cards.length, 6),
      );
      append_rest_to_deck_bytes(deckBytes, grouped_cards.length, 6);
    }
    let prev_card_number = 0;
    for (const card of grouped_cards) {
      const card_set_number = card.number.split("-")[1];
      const card_number_offset = card_set_number - prev_card_number;
      if (version == 0) {
        // 2 bits for card count (1-4)
        // 3 bits for parallel id (0-7)
        // 3 bits for start of card number offset
        deckBytes.push(
          ((card.count - 1) << 6) |
            (card.parallel_id << 3) |
            bits_with_carry(card_number_offset, 3),
        );
        // rest of card number offset
        append_rest_to_deck_bytes(deckBytes, card_number_offset, 3);
      } else {
        // 1 byte for card count (1-50 with BT6-085)
        // 3 bits for parallel id (0-7)
        // 5 bits for start of card number offset
        deckBytes.push(card.count - 1);
        deckBytes.push(
          (card.parallel_id << 5) | bits_with_carry(card_number_offset, 5),
        );
        // rest of card number offset
        append_rest_to_deck_bytes(deckBytes, card_number_offset, 5);
      }
      prev_card_number = card_set_number;
    }
  }

  // Compute and store cards checksum (second byte in buffer)
  // Only store the first byte of checksum
  deckBytes[1] = computeChecksum(deckBytes.slice(HEADER_SIZE));

  for (const b of deck_name_bytes) {
    deckBytes.push(b);
  }

  const deckString =
    PREFIX +
    (await bufferToBase64(new Uint8Array(deckBytes)))
      .replace(/\//g, "_")
      .replace(/\+/g, "-")
      .replace(/=/g, "");

  return deckString;
}
